src/util-file-decompression.c

```diff
@@ -0,0 +1,207 @@
+ * \param compress_depth how much compressed data we want to decompress
+ *
+ * \retval 1 if decompression works
+ * \retval 0 an error occured, and event set
+ */
+int FileDecompressFlashFile(const uint8_t *buffer, uint32_t buffer_len,
+                            uint8_t **decompressed_buffer,
+                            uint32_t *decompressed_buffer_len,
+                            DetectEngineThreadCtx *det_ctx,
+                            int swf_type,
+                            uint32_t decompress_depth,
+                            uint32_t compress_depth)
+{
+    int r = 0;
+    uint8_t *compressed_data = NULL;
+    uint8_t *decompressed_data = *decompressed_buffer;
+
+    int compression_type = FileIsFlashFile(buffer, buffer_len);
+    if (compression_type == FILE_FLASH_NO_COMPRESSION) {
+        return 0;
+    }
+
+    uint32_t offset = 0;
+    if (compression_type == FILE_FLASH_ZLIB_COMPRESSION) {
+        /* compressed data start from the 4th bytes */
+        offset = 8;
+    } else if (compression_type == FILE_FLASH_LZMA_COMPRESSION) {
+        /* compressed data start from the 17th bytes */
+        offset = 17;
+    }
+
+    uint32_t compressed_swf_len = 0;
+    if (buffer_len > offset && compress_depth == 0) {
+        compressed_swf_len = buffer_len - offset;
+    } else if (compress_depth > 0 && compress_depth <= buffer_len) {
+        compressed_swf_len = compress_depth;
+    } else if (compress_depth > 0 && compress_depth > buffer_len) {
+        compressed_swf_len = buffer_len;
+    } else {
+        compressed_swf_len = offset;
+    }
+
+    /* if compress_depth is 0, keep the buffer length */
+    uint32_t compressed_data_len = compressed_swf_len;
+
+    /* get flash decompressed file length */
+    uint32_t decompressed_swf_len = FileGetFlashDecompressedLen(buffer, buffer_len);
+    if (decompressed_swf_len == 0) {
+        decompressed_swf_len = MIN_SWF_LEN;
+    }
+
+    /* if decompress_depth is 0, keep the flash file length */
+    uint32_t decompressed_data_len = (decompress_depth == 0) ? decompressed_swf_len : decompress_depth;
+    decompressed_data = SCMalloc(decompressed_data_len + 8);
+    if (decompressed_data == NULL) {
+        DetectEngineSetEvent(det_ctx, FILE_DECODER_EVENT_NO_MEM);
+        return 0;
+    }
+
+    /* get file flash version */
+    uint8_t flash_version = FileGetFlashVersion(buffer, buffer_len);
+
+    /*
+     * FWS format
+     * | 4 bytes         | 4 bytes    | n bytes |
+     * | 'FWS' + version | script len | data    |
+     */
+    decompressed_data[0] = 'F';
+    decompressed_data[1] = 'W';
+    decompressed_data[2] = 'S';
+    decompressed_data[3] = flash_version;
+    memcpy(decompressed_data + 4, &decompressed_swf_len, 4);
+
+    if ((swf_type == HTTP_DECOMP_FLASH_ZLIB || swf_type == HTTP_DECOMP_FLASH_BOTH) &&
+        compression_type == FILE_FLASH_ZLIB_COMPRESSION)
+    {
+        if (flash_version < FLASH_ZLIB_MIN_VERSION) {
```

Comment:
Seems this check can be done before the memory allocation, making error handling simpler.